### PR TITLE
Update Dependencies for Python3.10

### PR DESCRIPTION
**Pull request by hydrobeam**
## Overview: What does this pull request change?
Changes a line in `pyproject.toml` to 
`python = ">=3.7,<3.11"`
from:
`python = ^3.7`

And then ran `poetry update`. 

## Motivation and Explanation: Why and how do your changes improve the library?

I tried running manim on python 3.10 with poetry and noticed that it was only failing on the scipy install, particularly in that it was installing `scipy 1.63`, not the latest scipy, which has support and binary wheels for python 3.10. I first tried to change this by doing `python = ">=3.7"`

But then got this error:
```python
  The current project's Python requirement (>=3.7) is not compatible with some of the required packages Python requirement:
    - flake8-pytest-style requires Python >=3.6,<4.0, so it will not be satisfied for Python >=4.0
    - flake8-pytest-style requires Python >=3.6,<4.0, so it will not be satisfied for Python >=4.0

  Because no versions of flake8-pytest-style match >1.5.0,<1.5.1 || >1.5.1,<2.0.0
   and flake8-pytest-style (1.5.0) requires Python >=3.6,<4.0, flake8-pytest-style is forbidden.
  So, because flake8-pytest-style (1.5.1) requires Python >=3.6,<4.0
   and manim depends on flake8-pytest-style (^1.5.0), version solving failed.
```

So i just specified that it must be less than python 3.11 and then ran `poetry update` and scipy was updated in the process. Then, I could successfully run a manim scene.

## Further Information and Comments

I'm not sure what the `^` symbol means, and spent a long time trying to figure out to no avail. I don't think it's particularly important, but I could be wrong.

## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by naveen521kk** [comment]
> I'm not sure what the ^ symbol means, and spent a long time trying to figure out to no avail. I don't think it's particularly important, but I could be wrong.

See https://python-poetry.org/docs/dependency-specification/#caret-requirements. `^3.7` means the same as `>=3.7,<4.0`. I'm not sure why poetry error's out though.

**Comment by marcin-serwin** [comment]
I've also encountered problems with this, and I think I understand why the change is important, so I'll describe it here for the sake of completeness:

- `scipy 1.6.1` specifies `python >=3.7` as its requirement, but it *does not* work with `python 3.10`
- `scipy 1.7.3` specifies `python >=3.7,<3.11` as its requirement, and it does work with `python 3.10`

This means that with `python = ^3.7` requirement in the project `poetry` cannot install newer `scipy` versions, because `1.6.1` is the last version that claims to work with arbitrarily high python versions. Thus:

- requirement `python = ^3.7` is able to satisfy all requirements but fails to install due to `scipy 1.6.1` incorrectly states that it works with `python 3.10`
- requirement `python >=3.7` makes it impossible to satisfy all requirements because `flake8-pytest-style` requires `python <= 4.0`. This is the error described in motivation.
- finally, requirement `python >=3.7,<3.11` is able to satisfy all requirements and allows installing the newest version of `scipy` that works correctly with `python 3.10`.

Hope this helps.

**Comment by christopher-besch** [comment]
Apparently `steps.cache-vars.outputs.poetry-venv-dir` isn't defined. I guess `poetry config virtualenvs.path` isn't returning anything.
The weird thing is that it seems to be working on my machine...
image


Ah, found it:
image
Just scroll up a bit. Apparently `cleo` isn't part of Python3.10 anymore.
Why did that one not crash the CI? We should look into this.

**Comment by christopher-besch** [comment]
Ok, looks like Poetry is using a [new installer](https://python-poetry.org/docs/master/#installing-with-the-official-installer). I've removed all other dependencies and added that installer.

**Comment by christopher-besch** [comment]
I think I've fixed your problem. But right now the installation of Manim fails for Python3.10.

**Comment by marcin-serwin** [comment]
The build fails, because `poetry` is constrained to old versions of `grpcio`:
https://github.com/ManimCommunity/manim/blob/7c7bd2b120c2fdf840b0fd333c651bbebf559d9c/pyproject.toml#L46-L47

Changing these lines to `^1.43.0` and running poetry update should fix the problem.

On a related note: the constraints with `*` will break sooner or later. Since the `pyproject.toml` is being updated anyway, we can also change these. So e.g.
https://github.com/ManimCommunity/manim/blob/7c7bd2b120c2fdf840b0fd333c651bbebf559d9c/pyproject.toml#L35
gets changed to
```
scipy = "^1.7.3"
```
where `1.7.3` is the newest version of `scipy` available on https://pypi.org/. `poetry` can do this automatically by running `poetry update scipy@latest`.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -36,14 +36,17 @@ jobs:
 
       - name: Install Poetry
         run: |
-          pip install wheel
-          curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python
+          pip install --user poetry
```
Comment (naveen521kk) [inline .github/workflows/ci.yml:39]: Thoughts about using `pipx` instead? I guess that would be better.
```suggestion
          pipx install poetry
```

File: pyproject.toml [hunk 2]
```diff
@@ -25,60 +26,61 @@ packages = [
...
 pycairo = "^1.19"
 manimpango = "^0.3.0"
 networkx = "^2.5"
 decorator = "^5.0.7"
-setuptools = "*"
-importlib-metadata = {version = "*", python = "<3.8"}
-grpcio =  { version = "1.33.*", optional = true }
-grpcio-tools = { version = "1.33.*", optional = true }
-watchdog = "*"
+setuptools = "^60.1.0"
```
Comment (naveen521kk) [inline pyproject.toml:44]: Do we really need `setuptools` to be installed? Can you point out where it is used?

File: pyproject.toml [hunk 3]
```diff
@@ -25,60 +26,61 @@ packages = [
...
+importlib-metadata = "^4.10.0"
```
Comment (naveen521kk) [inline pyproject.toml:45]: This dependency is only required for python 3.7.
So,
```suggestion
importlib-metadata = { version = "^4.10.0", python = "<3.8" }
```

File: .github/workflows/ci.yml (hunk 1 above)
Comment (christopher-besch) [reply .github/workflows/ci.yml:39]: I've spent an hour getting this stuff to work with all platforms. Don't touch it!

Comment (christopher-besch) [reply .github/workflows/ci.yml:39]: But honestly, I don't mind.

File: pyproject.toml [hunk 4]
```diff
@@ -25,60 +26,61 @@ packages = [
...
 mapbox-earcut = "^0.12.10"
 cloup = "^0.7.0"
-requests = "*"
+requests = "^2.26.0"
 dearpygui = { version = "^0.8", optional = true }
 screeninfo = "^0.6.7"
 skia-pathops = "^0.7.0"
 isosurfaces = "0.1.0"
 srt = "^3.5.0"
+Pygments = "^2.10.0"
```
Comment (naveen521kk) [inline pyproject.toml:60]: This is repeated twice. Remove either one of them.

File: pyproject.toml (hunk 2 above)
Comment (christopher-besch) [reply pyproject.toml:44]: I have no idea, I didn't add it; I only fixed the `*` notation.

File: pyproject.toml (hunk 4 above)
Comment (christopher-besch) [reply pyproject.toml:60]: Poetry did a bad there...

**Comment by christopher-besch** [comment]
Now `psutil` is causing problems on Windows.

**Comment by marcin-serwin** [comment]
> Now `psutil` is causing problems on Windows.

This is probably related to this: https://github.com/giampaolo/psutil/issues/2028.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (naveen521kk) [reply .github/workflows/ci.yml:39]: > I've spent an hour getting this stuff to work with all platforms. Don't touch it!

You could have asked me beforehand :D

File: pyproject.toml (hunk 2 above)
Comment (naveen521kk) [reply pyproject.toml:43]: In that case, can we remove it?

File: pyproject.toml [hunk 5]
```diff
@@ -87,6 +87,7 @@ flake8-rst-docstrings = "^0.2.3"
 flake8-pytest-style = "^1.5.0"
 flake8-simplify = "^0.14.1"
 flake8-comprehensions = "^3.7.0"
+psutil-wheels = "^5.8.1"
```
Comment (marcin-serwin) [inline pyproject.toml:90]: This should replace `psutil`, not be an additional package.

Comment (marcin-serwin) [reply pyproject.toml:90]: Also use `^5.8.0`, since `5.8.1` has been yanked.

File: pyproject.toml [hunk 6]
```diff
@@ -25,67 +26,67 @@ packages = [
...
 pytest-xdist = "^2.2"
-psutil = ">=3.0"
 flake8-builtins = "^1.5.3"
 flake8-bugbear = "^21.4.3"
 flake8-docstrings = "^1.6.0"
 flake8-rst-docstrings = "^0.2.3"
 flake8-pytest-style = "^1.5.0"
 flake8-simplify = "^0.14.1"
 flake8-comprehensions = "^3.7.0"
+psutil-wheels = "^5.8.1"
```
Comment (marcin-serwin) [inline pyproject.toml:89]: Poetry doesn't support yanked releases, https://github.com/python-poetry/poetry/issues/2453.
```suggestion
psutil-wheels = "5.8.0"
```

**Comment by marcin-serwin** [comment]
`poetry update` needs to be run when changing dependencies.

**Comment by christopher-besch** [comment]
I'm getting old...

File: pyproject.toml [hunk 7]
```diff
@@ -25,67 +26,67 @@ packages = [
...
 pytest-xdist = "^2.2"
-psutil = ">=3.0"
 flake8-builtins = "^1.5.3"
 flake8-bugbear = "^21.4.3"
 flake8-docstrings = "^1.6.0"
 flake8-rst-docstrings = "^0.2.3"
 flake8-pytest-style = "^1.5.0"
 flake8-simplify = "^0.14.1"
 flake8-comprehensions = "^3.7.0"
+psutil-wheels = "5.8.0"
```
Comment (marcin-serwin) [inline pyproject.toml:89]: Eh, I didn't notice that `psutil-wheels` doesn't support `python 3.7`. In that case this needs to be changed to
```suggestion
psutil-wheels = {version = "5.8.0", python = ">=3.10"}
psutil = {version = "^5.8.0", python = "<3.10"}
```
And of course `poetry update` :>

Comment (christopher-besch) [reply pyproject.toml:89]: That `Commit suggestion` button is just too seductive xD

**Comment by christopher-besch** [comment]
Finally, thanks a lot @marcin-serwin!

**Review by christopher-besch — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -36,14 +36,17 @@ jobs:
...
+
+      - name: Setup macOS PATH
+        if: runner.os == 'macOS'
+        run: |
+          echo "$HOME/.local/bin" >> $GITHUB_PATH
```
Comment (naveen521kk) [inline .github/workflows/ci.yml:44]: Why this is needed separately? I think you can simply change the below line 
https://github.com/ManimCommunity/manim/blob/be6c35f8a62671a07fc0d4226c3108f52f8bc04f/.github/workflows/ci.yml#L107
to match this change.

**Review by naveen521kk — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (naveen521kk) [reply pyproject.toml:43]: Does manim fail to run without this dependency?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (christopher-besch) [reply .github/workflows/ci.yml:44]: Nope, the problem is that the very next job, `Setup cache variables` requires `poetry` to be available.

Comment (naveen521kk) [reply .github/workflows/ci.yml:44]: > Nope, the problem is that the very next job, Setup cache variables requires poetry to be available.

I see but why not move that block of code here or just add to the path temporarily as before?

Comment (christopher-besch) [reply .github/workflows/ci.yml:44]: I thought about doing that, but I figured we had some reason for putting the dependency installation after setting up the caches.

**Comment by Darylgolden** [comment]
After resolving the merge conflict, will the PR be ready?

**Comment by christopher-besch** [comment]
What the hell happened to the history, did I screw this up?

**Comment by naveen521kk** [comment]
> What the hell happened to the history, did I screw this up?

I fixed it by clicking a single button 😄

**Comment by christopher-besch** [comment]
> > What the hell happened to the history, did I screw this up?
> 
> I fixed it by clicking a single button smile

What'd you do?

**Comment by naveen521kk** [comment]
> Which button?

Update branch button which GitHub shows.

> And thanks for killing the CI, I totally wasn't anxiously waiting for the result of whether I managed to finally clean up this dependency mess. Ahhhh ;)

Sorry!

**Review by Darylgolden — APPROVED** [review]
LGTM

File: pyproject.toml [hunk 9]
```diff
@@ -25,61 +26,60 @@ packages = [
...
+grpcio =  { version = "^1.43.0", optional = true }
+grpcio-tools = { version = "^1.43.0", optional = true }
+watchdog = "^2.1.6"
 jupyterlab = { version = "^3.0", optional = true }
 moderngl = "^5.6.3"
 moderngl-window = "^2.3.0"
 mapbox-earcut = "^0.12.10"
 cloup = "^0.7.0"
-requests = "*"
+requests = "^2.26.0"
```
Comment (naveen521kk) [inline pyproject.toml:53]: Pin only the major version? I don't think requests will break API between minor releases.
```suggestion
requests = "^2"
```

Comment (christopher-besch) [reply pyproject.toml:53]: Shouldn't it then be '"^2.26"'?

Comment (marcin-serwin) [reply pyproject.toml:53]: [Caret requirements](https://python-poetry.org/docs/dependency-specification/#caret-requirements) only pin major version. This means that
- `^2` accepts >= `2.0.0`, < `3.0.0`
- `^2.26` accepts >= `2.26.0`, < `3.0.0`
- `^2.26.2` accepts >= `2.26.2`, < `3.0.0`
I think it should stay the way it is currently.

Comment (naveen521kk) [reply pyproject.toml:53]: > I think it should stay the way it is currently.

I mean even older versions (from `2.0.0`) should work right?

Comment (marcin-serwin) [reply pyproject.toml:53]: Not necessarily. Manim can rely on some features introduced in later versions (e.g. python 3.10 support). It'd have to be tested and advantages of relaxing constraints that way are minuscule.

Comment (Darylgolden) [reply pyproject.toml:53]: Also, it doesn't really make sense to relax the restriction here and not the others.

Comment (naveen521kk) [reply pyproject.toml:53]: > Not necessarily. Manim can rely on some features introduced in later versions (e.g. python 3.10 support). It'd have to be tested and advantages of relaxing constraints that way are minuscule.

Make sense.

**Review by naveen521kk — APPROVED** [review]
LGTM!

**Pull request merged**